docs/proposals/selinux-enhancements.md

```diff
@@ -0,0 +1,132 @@
+    directory with a label usable from a container, so that I can limit the
+    attack surface available to a container escaping its namespace isolation
+2.  As a user, I want to run a pod without an SELinux context explicitly
+    specified and be isolated using MCS (multi-category security) on systems
+    where SELinux is enabled, so that the pods on each host are isolated from
+    one another
+3.  As a user, I want to run a pod that uses the host IPC or PID namespace and
+    want the system to do the right thing with regard to SELinux, so that no
+    unnecessary relabel actions are performed
+
+### Labeling the Kubelet directory
+
+As previously stated, the current codebase relies on the Kubelet directory
+being labeled with an SELinux context usable from a container.  The Kubelet
+uses the SELinux context of this directory to determine what SELinux context
+`tmpfs` mounts (provided by the EmptyDir memory-medium option) should receive.
+The problem with this is that it opens an attack surface to a container that
+escapes its namespace isolation.
+
+### SELinux when no context is specified
+
+When no SELinux context is specified, Kubernetes should just do the right
+thing, where DTRT is defined as isolating pods with a node-unique set of
+categories.  Node-uniqueness means unique among the pods scheduled onto the
+node.  Long-term, we want to have a cluster-wide allocator for MCS labels.
+Node-unique MCS labels are a good middle ground that is possible without a
+new, large, feature.
+
+### SELinux and host IPC and PID namespaces
+
+Containers in pods that use the host IPC or PID namespaces are run with the
+`spc_t` label, which means that the container process is effectively
+unconfined by SELinux.  In this case, no relabeling is necessary.
+
+## Analysis
+
+### Libcontainer SELinux library
+
+Docker and rkt both use the libcontainer SELinux library.  This library
+provides a method, `GetLxcContexts`, that returns the correct base SELinux
+contexts for container processes and files used by them.  Docker and rkt both
+leverage this call to determine the 'starting' SELinux contexts for
+containers.
+
+### Docker
+
+Docker's behavior when no SELinux context is defined is to pick a unique MCS
+label for a container.  Additionally, a container B shares the IPC namespace
+of another container A, container B is given the SELinux context of container
+A.  What this means for Kubernetes pods (where containers share the IPC
+namespace of the infra-container) is that in a vacuum the containers in a pod
+should have the same SELinux context.
+
+Docker is capable of relabeling bind-mounts into containers using the `:Z`
+bind-mount flag.  However, in the current implementation of the docker runtime
+in Kubernetes, the `:Z` option is only applied when the pod's SecurityContext
+contains an SELinux context.  We could easily implement the correct behaviors
+by always setting `:Z` on systems where SELinux is enabled.
```

Comment:
Question: If container A and container B are in the same IPC namespace (infra container, and user container), and `:Z` (uppercase) is applied during mount. Are they sharing the same mcs label ?